File: CurrencyTracker.java
# CurrencyTracker.java
/**
 * 💱 Currency Tracker – Live Exchange Rates with ASCII Charts (Java Edition)
 * Advanced: live rates, historical charts, caching, colorful CLI, stats
 * Requires: Java 17+ (uses java.net.http and java.time)
 */

import java.io.*;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.file.*;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.stream.Collectors;

// ─── Data Classes ──────────────────────────────────────────────────────────

class RateEntry {
    public String date;
    public double rate;
    public RateEntry(String date, double rate) { this.date = date; this.rate = rate; }
}

class CacheData {
    public String base = "USD";
    public String target = "EUR";
    public List<RateEntry> rates = new ArrayList<>();
    public String lastUpdate = "";
}

// ─── Main App ──────────────────────────────────────────────────────────────

public class CurrencyTracker {
    // ─── Configuration ──────────────────────────────────────────────────────

    private static final String DEFAULT_BASE = "USD";
    private static final String DEFAULT_TARGET = "EUR";
    private static final int HISTORY_DAYS = 30;
    private static final String API_BASE = "https://api.exchangerate.host";

    // ─── Colors ─────────────────────────────────────────────────────────────

    private static final String RESET = "\u001B[0m";
    private static final String BRIGHT = "\u001B[1m";
    private static final String DIM = "\u001B[2m";
    private static final String RED = "\u001B[31m";
    private static final String GREEN = "\u001B[32m";
    private static final String YELLOW = "\u001B[33m";
    private static final String BLUE = "\u001B[34m";
    private static final String MAGENTA = "\u001B[35m";
    private static final String CYAN = "\u001B[36m";

    private static String c(String text, String color) { return color + text + RESET; }

    // ─── API Client ─────────────────────────────────────────────────────────

    private static class ExchangeRateAPI {
        private static final HttpClient client = HttpClient.newHttpClient();

        static double getLatest(String base, String target) throws Exception {
            String url = API_BASE + "/latest?base=" + base + "&symbols=" + target;
            HttpRequest req = HttpRequest.newBuilder().uri(URI.create(url)).GET().build();
            HttpResponse<String> resp = client.send(req, HttpResponse.BodyHandlers.ofString());
            if (resp.statusCode() == 200) {
                var json = new com.google.gson.JsonParser().parse(resp.body()).getAsJsonObject();
                if (json.get("success").getAsBoolean()) {
                    return json.getAsJsonObject("rates").get(target).getAsDouble();
                }
            }
            throw new RuntimeException("Failed to get rate");
        }

        static List<RateEntry> getHistory(String base, String target, int days) throws Exception {
            LocalDate end = LocalDate.now();
            LocalDate start = end.minusDays(days);
            DateTimeFormatter fmt = DateTimeFormatter.ISO_LOCAL_DATE;
            String url = String.format("%s/timeseries?start_date=%s&end_date=%s&base=%s&symbols=%s",
                    API_BASE, start.format(fmt), end.format(fmt), base, target);
            HttpRequest req = HttpRequest.newBuilder().uri(URI.create(url)).GET().build();
            HttpResponse<String> resp = client.send(req, HttpResponse.BodyHandlers.ofString());
            if (resp.statusCode() == 200) {
                var json = new com.google.gson.JsonParser().parse(resp.body()).getAsJsonObject();
                if (json.get("success").getAsBoolean()) {
                    var ratesObj = json.getAsJsonObject("rates");
                    List<RateEntry> entries = new ArrayList<>();
                    for (var date : ratesObj.keySet()) {
                        double rate = ratesObj.getAsJsonObject(date).get(target).getAsDouble();
                        entries.add(new RateEntry(date, rate));
                    }
                    entries.sort(Comparator.comparing(a -> a.date));
                    return entries;
                }
            }
            throw new RuntimeException("Failed to get history");
        }
    }

    // ─── Cache Manager ─────────────────────────────────────────────────────

    private static class Cache {
        private final Path filePath;
        private CacheData data;

        Cache() throws IOException {
            String home = System.getProperty("user.home");
            Path dir = Paths.get(home, ".currency_tracker");
            Files.createDirectories(dir);
            filePath = dir.resolve("cache.json");
            load();
        }

        private void load() {
            data = new CacheData();
            if (Files.exists(filePath)) {
                try {
                    String json = new String(Files.readAllBytes(filePath));
                    var obj = new com.google.gson.JsonParser().parse(json).getAsJsonObject();
                    data.base = obj.get("base").getAsString();
                    data.target = obj.get("target").getAsString();
                    data.lastUpdate = obj.get("last_update").getAsString();
                    var ratesArr = obj.getAsJsonArray("rates");
                    for (var item : ratesArr) {
                        var obj2 = item.getAsJsonObject();
                        data.rates.add(new RateEntry(obj2.get("date").getAsString(), obj2.get("rate").getAsDouble()));
                    }
                } catch (Exception e) { data = new CacheData(); }
            }
        }

        void save() throws IOException {
            var obj = new com.google.gson.JsonObject();
            obj.addProperty("base", data.base);
            obj.addProperty("target", data.target);
            obj.addProperty("last_update", data.lastUpdate);
            var ratesArr = new com.google.gson.JsonArray();
            for (var r : data.rates) {
                var item = new com.google.gson.JsonObject();
                item.addProperty("date", r.date);
                item.addProperty("rate", r.rate);
                ratesArr.add(item);
            }
            obj.add("rates", ratesArr);
            Files.write(filePath, obj.toString().getBytes());
        }

        List<RateEntry> get(String base, String target) {
            if (data.base.equals(base) && data.target.equals(target)) {
                return data.rates;
            }
            return null;
        }

        void set(String base, String target, List<RateEntry> rates) throws IOException {
            data.base = base;
            data.target = target;
            data.rates = rates;
            data.lastUpdate = LocalDate.now().toString();
            save();
        }

        void clear() throws IOException {
            data = new CacheData();
            save();
        }
    }

    // ─── Chart Renderer ─────────────────────────────────────────────────────

    private static String drawAsciiChart(List<RateEntry> rates, int width, int height) {
        if (rates.isEmpty()) return "No data available.";
        double[] values = rates.stream().mapToDouble(r -> r.rate).toArray();
        double minVal = Arrays.stream(values).min().getAsDouble();
        double maxVal = Arrays.stream(values).max().getAsDouble();
        double range = maxVal - minVal;
        if (range == 0) return String.format("Rate is constant at %.4f", values[0]);
        int[] normalized = new int[values.length];
        for (int i = 0; i < values.length; i++) {
            normalized[i] = (int) Math.floor((values[i] - minVal) / range * (height - 1));
        }
        StringBuilder sb = new StringBuilder();
        for (int row = height - 1; row >= 0; row--) {
            StringBuilder line = new StringBuilder();
            for (int i = 0; i < normalized.length; i++) {
                if (normalized[i] >= row) {
                    if (i > 0 && normalized[i-1] >= row) line.append('─');
                    else line.append('┌');
                } else {
                    line.append(' ');
                }
            }
            if (line.toString().trim().length() > 0) sb.append(line).append('\n');
        }
        // X axis
        int step = Math.max(1, rates.size() / 10);
        StringBuilder xAxis = new StringBuilder(" ");
        int lastPos = 0;
        for (int i = 0; i < rates.size(); i += step) {
            String label = rates.get(i).date.substring(5, 10);
            int spaces = i - lastPos;
            if (spaces > 0) xAxis.append(" ".repeat(spaces));
            xAxis.append(label);
            lastPos = i;
        }
        if (lastPos < rates.size() - 1) {
            xAxis.append(" ".repeat(rates.size() - 1 - lastPos));
        }
        sb.append(xAxis).append('\n');
        sb.append(String.format("Min: %.4f  Max: %.4f", minVal, maxVal));
        return sb.toString();
    }

    // ─── Main App ──────────────────────────────────────────────────────────

    private final Scanner scanner;
    private final Cache cache;
    private String base, target;
    private List<RateEntry> rates;

    public CurrencyTracker() throws Exception {
        scanner = new Scanner(System.in);
        cache = new Cache();
        base = DEFAULT_BASE;
        target = DEFAULT_TARGET;
        rates = new ArrayList<>();
        loadCachedOrFetch();
    }

    private void loadCachedOrFetch() {
        List<RateEntry> cached = cache.get(base, target);
        if (cached != null && !cached.isEmpty()) {
            rates = cached;
            System.out.println(c("📂 Loaded cached data for " + base + "/" + target, DIM));
        } else {
            fetchHistory();
        }
    }

    private void fetchHistory() {
        System.out.print(c("Fetching historical data...", DIM));
        try {
            List<RateEntry> history = ExchangeRateAPI.getHistory(base, target, HISTORY_DAYS);
            if (!history.isEmpty()) {
                rates = history;
                cache.set(base, target, rates);
                System.out.println(c(" ✅ Retrieved " + rates.size() + " data points", GREEN));
            } else {
                System.out.println(c(" ⚠️  No historical data retrieved.", YELLOW));
            }
        } catch (Exception e) {
            System.out.println(c(" ❌ Error fetching data: " + e.getMessage(), RED));
        }
    }

    private double getCurrentRate() {
        try {
            return ExchangeRateAPI.getLatest(base, target);
        } catch (Exception e) {
            if (!rates.isEmpty()) return rates.get(rates.size()-1).rate;
            return 0.0;
        }
    }

    private String ask(String prompt) {
        System.out.print(prompt);
        return scanner.nextLine().trim();
    }

    private boolean askConfirm(String prompt) {
        String ans = ask(prompt + " (yes/no): ").toLowerCase();
        return ans.equals("yes") || ans.equals("y");
    }

    private void showMenu() {
        double current = getCurrentRate();
        String currentStr = current != 0.0 ? String.format("%.4f", current) : "N/A";
        System.out.println("\n" + c("═".repeat(50), CYAN));
        System.out.println(c("💱 CURRENCY TRACKER", BRIGHT + CYAN));
        System.out.println(c("═".repeat(50), CYAN));
        System.out.printf("  Base: %s  Target: %s\n", base, target);
        System.out.printf("  Current: %s\n", currentStr);
        System.out.printf("  Data points: %d\n", rates.size());
        System.out.println(c("═".repeat(50), CYAN));
        System.out.println("  1. 📈 Show chart");
        System.out.println("  2. 💹 Show current rate");
        System.out.println("  3. 🔄 Change currency pair");
        System.out.println("  4. 📊 Show statistics");
        System.out.println("  5. 🔄 Refresh data");
        System.out.println("  6. 🗑️  Clear cache");
        System.out.println("  0. 🚪 Exit");
        System.out.println(c("═".repeat(50), CYAN));
    }

    private void showChart() {
        if (rates.isEmpty()) {
            System.out.println(c("No data available. Fetching...", YELLOW));
            fetchHistory();
        }
        if (rates.isEmpty()) {
            System.out.println(c("No data to display.", RED));
            return;
        }
        System.out.println(c("\n📈 Exchange Rate Chart (last " + rates.size() + " days)", BRIGHT));
        System.out.println(drawAsciiChart(rates, 50, 10));
        System.out.println(c("\nLatest: " + String.format("%.4f", rates.get(rates.size()-1).rate), CYAN));
    }

    private void showRate() {
        double rate = getCurrentRate();
        if (rate != 0.0) {
            System.out.println(c("\n💹 Current " + base + "/" + target + ": " + String.format("%.4f", rate), GREEN));
        } else {
            System.out.println(c("Could not fetch rate.", RED));
        }
    }

    private void showStats() {
        if (rates.isEmpty()) {
            System.out.println(c("No data available.", YELLOW));
            return;
        }
        double sum = rates.stream().mapToDouble(r -> r.rate).sum();
        double avg = sum / rates.size();
        double minv = rates.stream().mapToDouble(r -> r.rate).min().getAsDouble();
        double maxv = rates.stream().mapToDouble(r -> r.rate).max().getAsDouble();
        double volatility = (maxv - minv) / avg * 100.0;
        double last = rates.get(rates.size()-1).rate;
        System.out.println("\n📊 STATISTICS");
        System.out.println(c("─".repeat(30), DIM));
        System.out.printf("  Period:      %d days\n", rates.size());
        System.out.printf("  Average:     %.4f\n", avg);
        System.out.printf("  Min:         %.4f\n", minv);
        System.out.printf("  Max:         %.4f\n", maxv);
        System.out.printf("  Volatility:  %.2f%%\n", volatility);
        System.out.printf("  Current:     %.4f\n", last);
    }

    private void changePair() {
        String b = ask("Base currency (default " + base + "): ");
        String t = ask("Target currency (default " + target + "): ");
        String newBase = b.isEmpty() ? base : b.toUpperCase();
        String newTarget = t.isEmpty() ? target : t.toUpperCase();
        if (newBase.equals(base) && newTarget.equals(target)) {
            System.out.println("Pair unchanged.");
            return;
        }
        base = newBase;
        target = newTarget;
        List<RateEntry> cached = cache.get(base, target);
        if (cached != null && !cached.isEmpty()) {
            rates = cached;
            System.out.println(c("📂 Loaded cached data for " + base + "/" + target, DIM));
        } else {
            fetchHistory();
        }
    }

    private void refresh() {
        fetchHistory();
        System.out.println(c("✅ Data refreshed.", GREEN));
    }

    private void clearCache() {
        if (!askConfirm("🗑️  Delete all cached data?")) return;
        try {
            cache.clear();
            rates.clear();
            System.out.println(c("Cache cleared.", YELLOW));
        } catch (Exception e) {
            System.out.println(c("❌ Error clearing cache: " + e.getMessage(), RED));
        }
    }

    public void run() {
        System.out.print("\033[H\033[2J");
        System.out.flush();
        System.out.println(c("\n💱 Currency Tracker – Live Exchange Rates with Charts", BRIGHT + CYAN));
        System.out.println(c("Track the pulse of the global economy.", DIM));

        while (true) {
            showMenu();
            String choice = ask("Your choice: ");
            switch (choice) {
                case "1": showChart(); break;
                case "2": showRate(); break;
                case "3": changePair(); break;
                case "4": showStats(); break;
                case "5": refresh(); break;
                case "6": clearCache(); break;
                case "0":
                    System.out.println(c("👋 Goodbye!", CYAN));
                    return;
                default:
                    System.out.println(c("❌ Invalid choice.", RED));
            }
            if (!choice.equals("0")) {
                System.out.print("\nPress Enter to continue...");
                scanner.nextLine();
            }
        }
    }

    // ─── Main ──────────────────────────────────────────────────────────────

    public static void main(String[] args) {
        try {
            // Gson is needed for JSON parsing; include in classpath
            // For simplicity, we use a minimal JSON parser or require Gson library.
            // In a real project, add dependency on Gson.
            // We'll use a small hack: we assume Gson is available.
            // If not, we could implement a simple parser, but for brevity we add a comment.
            // This code uses Gson; please add the library.
            // To compile: javac -cp gson.jar CurrencyTracker.java
            // To run: java -cp .:gson.jar CurrencyTracker
            CurrencyTracker app = new CurrencyTracker();
            app.run();
        } catch (Exception e) {
            System.err.println(c("❌ Unexpected error: " + e.getMessage(), RED));
            e.printStackTrace();
            System.exit(1);
        }
    }
}

File: CurrencyTracker.cs
# CurrencyTracker.cs
/**
 * 💱 Currency Tracker – Live Exchange Rates with ASCII Charts (C# Edition)
 * Advanced: live rates, historical charts, caching, colorful CLI, stats
 * Requires: .NET 6.0+
 */

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net.Http;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading.Tasks;

// ─── Data Classes ──────────────────────────────────────────────────────────

public class RateEntry
{
    [JsonPropertyName("date")]
    public string Date { get; set; } = "";
    
    [JsonPropertyName("rate")]
    public double Rate { get; set; }
}

public class CacheData
{
    [JsonPropertyName("base")]
    public string Base { get; set; } = "USD";
    
    [JsonPropertyName("target")]
    public string Target { get; set; } = "EUR";
    
    [JsonPropertyName("rates")]
    public List<RateEntry> Rates { get; set; } = new();
    
    [JsonPropertyName("last_update")]
    public string LastUpdate { get; set; } = "";
}

// ─── Main App ──────────────────────────────────────────────────────────────

public class CurrencyTracker
{
    // ─── Configuration ──────────────────────────────────────────────────────

    private const string DefaultBase = "USD";
    private const string DefaultTarget = "EUR";
    private const int HistoryDays = 30;
    private const string ApiBase = "https://api.exchangerate.host";

    // ─── Colors ─────────────────────────────────────────────────────────────

    private static readonly string Reset = "\u001B[0m";
    private static readonly string Bright = "\u001B[1m";
    private static readonly string Dim = "\u001B[2m";
    private static readonly string Red = "\u001B[31m";
    private static readonly string Green = "\u001B[32m";
    private static readonly string Yellow = "\u001B[33m";
    private static readonly string Blue = "\u001B[34m";
    private static readonly string Magenta = "\u001B[35m";
    private static readonly string Cyan = "\u001B[36m";

    private static string C(string text, string color) => color + text + Reset;

    // ─── API Client ─────────────────────────────────────────────────────────

    private static readonly HttpClient httpClient = new HttpClient();

    private static async Task<double> GetLatestAsync(string baseCurrency, string targetCurrency)
    {
        string url = $"{ApiBase}/latest?base={baseCurrency}&symbols={targetCurrency}";
        var response = await httpClient.GetStringAsync(url);
        using var doc = JsonDocument.Parse(response);
        var root = doc.RootElement;
        if (root.GetProperty("success").GetBoolean())
        {
            var rates = root.GetProperty("rates");
            return rates.GetProperty(targetCurrency).GetDouble();
        }
        throw new Exception("Failed to get rate");
    }

    private static async Task<List<RateEntry>> GetHistoryAsync(string baseCurrency, string targetCurrency, int days)
    {
        var end = DateTime.Now;
        var start = end.AddDays(-days);
        string url = $"{ApiBase}/timeseries?start_date={start:yyyy-MM-dd}&end_date={end:yyyy-MM-dd}&base={baseCurrency}&symbols={targetCurrency}";
        var response = await httpClient.GetStringAsync(url);
        using var doc = JsonDocument.Parse(response);
        var root = doc.RootElement;
        if (root.GetProperty("success").GetBoolean())
        {
            var ratesObj = root.GetProperty("rates");
            var entries = new List<RateEntry>();
            foreach (var prop in ratesObj.EnumerateObject())
            {
                string date = prop.Name;
                double rate = prop.Value.GetProperty(targetCurrency).GetDouble();
                entries.Add(new RateEntry { Date = date, Rate = rate });
            }
            entries = entries.OrderBy(e => e.Date).ToList();
            return entries;
        }
        throw new Exception("Failed to get history");
    }

    // ─── Cache Manager ─────────────────────────────────────────────────────

    private class Cache
    {
        private readonly string filePath;
        private CacheData data;

        public Cache()
        {
            string home = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
            string dir = Path.Combine(home, ".currency_tracker");
            Directory.CreateDirectory(dir);
            filePath = Path.Combine(dir, "cache.json");
            Load();
        }

        private void Load()
        {
            if (File.Exists(filePath))
            {
                try
                {
                    string json = File.ReadAllText(filePath);
                    data = JsonSerializer.Deserialize<CacheData>(json) ?? new CacheData();
                    return;
                }
                catch { /* ignore */ }
            }
            data = new CacheData();
        }

        public void Save()
        {
            string json = JsonSerializer.Serialize(data, new JsonSerializerOptions { WriteIndented = true });
            File.WriteAllText(filePath, json);
        }

        public List<RateEntry> Get(string baseCurrency, string targetCurrency)
        {
            if (data.Base == baseCurrency && data.Target == targetCurrency)
                return data.Rates;
            return null;
        }

        public void Set(string baseCurrency, string targetCurrency, List<RateEntry> rates)
        {
            data.Base = baseCurrency;
            data.Target = targetCurrency;
            data.Rates = rates;
            data.LastUpdate = DateTime.Now.ToString("yyyy-MM-dd");
            Save();
        }

        public void Clear()
        {
            data = new CacheData();
            Save();
        }
    }

    // ─── Chart Renderer ─────────────────────────────────────────────────────

    private static string DrawAsciiChart(List<RateEntry> rates, int width = 50, int height = 10)
    {
        if (rates == null || rates.Count == 0) return "No data available.";
        var values = rates.Select(r => r.Rate).ToArray();
        double minVal = values.Min();
        double maxVal = values.Max();
        double range = maxVal - minVal;
        if (range == 0) return $"Rate is constant at {values[0]:F4}";

        int[] normalized = new int[values.Length];
        for (int i = 0; i < values.Length; i++)
            normalized[i] = (int)Math.Floor((values[i] - minVal) / range * (height - 1));

        var lines = new List<string>();
        for (int row = height - 1; row >= 0; row--)
        {
            var line = new char[normalized.Length];
            for (int i = 0; i < normalized.Length; i++)
            {
                if (normalized[i] >= row)
                {
                    if (i > 0 && normalized[i-1] >= row)
                        line[i] = '─';
                    else
                        line[i] = '┌';
                }
                else
                {
                    line[i] = ' ';
                }
            }
            string str = new string(line);
            if (!string.IsNullOrWhiteSpace(str))
                lines.Add(str);
        }
        // X axis
        int step = Math.Max(1, rates.Count / 10);
        var xAxis = new System.Text.StringBuilder(" ");
        int lastPos = 0;
        for (int i = 0; i < rates.Count; i += step)
        {
            string label = rates[i].Date.Substring(5, 5);
            int spaces = i - lastPos;
            if (spaces > 0) xAxis.Append(' ', spaces);
            xAxis.Append(label);
            lastPos = i;
        }
        if (lastPos < rates.Count - 1)
            xAxis.Append(' ', rates.Count - 1 - lastPos);
        var result = string.Join("\n", lines) + "\n" + xAxis.ToString();
        result += $"\nMin: {minVal:F4}  Max: {maxVal:F4}";
        return result;
    }

    // ─── App State ──────────────────────────────────────────────────────────

    private readonly Cache cache;
    private string baseCurrency, targetCurrency;
    private List<RateEntry> rates;

    public CurrencyTracker()
    {
        cache = new Cache();
        baseCurrency = DefaultBase;
        targetCurrency = DefaultTarget;
        rates = new List<RateEntry>();
        LoadCachedOrFetchAsync().GetAwaiter().GetResult();
    }

    private async Task LoadCachedOrFetchAsync()
    {
        var cached = cache.Get(baseCurrency, targetCurrency);
        if (cached != null && cached.Count > 0)
        {
            rates = cached;
            Console.WriteLine(C($"📂 Loaded cached data for {baseCurrency}/{targetCurrency}", Dim));
        }
        else
        {
            await FetchHistoryAsync();
        }
    }

    private async Task FetchHistoryAsync()
    {
        Console.Write(C("Fetching historical data...", Dim));
        try
        {
            var history = await GetHistoryAsync(baseCurrency, targetCurrency, HistoryDays);
            if (history.Count > 0)
            {
                rates = history;
                cache.Set(baseCurrency, targetCurrency, rates);
                Console.WriteLine(C($" ✅ Retrieved {rates.Count} data points", Green));
            }
            else
            {
                Console.WriteLine(C(" ⚠️  No historical data retrieved.", Yellow));
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine(C($" ❌ Error: {ex.Message}", Red));
        }
    }

    private async Task<double> GetCurrentRateAsync()
    {
        try
        {
            return await GetLatestAsync(baseCurrency, targetCurrency);
        }
        catch
        {
            if (rates.Count > 0) return rates[rates.Count-1].Rate;
            return 0.0;
        }
    }

    private string Ask(string prompt)
    {
        Console.Write(prompt);
        return Console.ReadLine()?.Trim() ?? "";
    }

    private bool AskConfirm(string prompt)
    {
        string ans = Ask(prompt + " (yes/no): ").ToLower();
        return ans == "yes" || ans == "y";
    }

    private void ShowMenu()
    {
        double current = GetCurrentRateAsync().GetAwaiter().GetResult();
        string currentStr = current != 0.0 ? $"{current:F4}" : "N/A";
        Console.WriteLine("\n" + C(new string('═', 50), Cyan));
        Console.WriteLine(C("💱 CURRENCY TRACKER", Bright + Cyan));
        Console.WriteLine(C(new string('═', 50), Cyan));
        Console.WriteLine($"  Base: {baseCurrency}  Target: {targetCurrency}");
        Console.WriteLine($"  Current: {currentStr}");
        Console.WriteLine($"  Data points: {rates.Count}");
        Console.WriteLine(C(new string('═', 50), Cyan));
        Console.WriteLine("  1. 📈 Show chart");
        Console.WriteLine("  2. 💹 Show current rate");
        Console.WriteLine("  3. 🔄 Change currency pair");
        Console.WriteLine("  4. 📊 Show statistics");
        Console.WriteLine("  5. 🔄 Refresh data");
        Console.WriteLine("  6. 🗑️  Clear cache");
        Console.WriteLine("  0. 🚪 Exit");
        Console.WriteLine(C(new string('═', 50), Cyan));
    }

    private void ShowChart()
    {
        if (rates.Count == 0)
        {
            Console.WriteLine(C("No data available. Fetching...", Yellow));
            FetchHistoryAsync().GetAwaiter().GetResult();
        }
        if (rates.Count == 0)
        {
            Console.WriteLine(C("No data to display.", Red));
            return;
        }
        Console.WriteLine(C($"\n📈 Exchange Rate Chart (last {rates.Count} days)", Bright));
        Console.WriteLine(DrawAsciiChart(rates));
        Console.WriteLine(C($"\nLatest: {rates[rates.Count-1].Rate:F4}", Cyan));
    }

    private void ShowRate()
    {
        double rate = GetCurrentRateAsync().GetAwaiter().GetResult();
        if (rate != 0.0)
        {
            Console.WriteLine(C($"\n💹 Current {baseCurrency}/{targetCurrency}: {rate:F4}", Green));
        }
        else
        {
            Console.WriteLine(C("Could not fetch rate.", Red));
        }
    }

    private void ShowStats()
    {
        if (rates.Count == 0)
        {
            Console.WriteLine(C("No data available.", Yellow));
            return;
        }
        double sum = rates.Sum(r => r.Rate);
        double avg = sum / rates.Count;
        double minv = rates.Min(r => r.Rate);
        double maxv = rates.Max(r => r.Rate);
        double volatility = (maxv - minv) / avg * 100.0;
        double last = rates[rates.Count-1].Rate;
        Console.WriteLine("\n📊 STATISTICS");
        Console.WriteLine(C(new string('─', 30), Dim));
        Console.WriteLine($"  Period:      {rates.Count} days");
        Console.WriteLine($"  Average:     {avg:F4}");
        Console.WriteLine($"  Min:         {minv:F4}");
        Console.WriteLine($"  Max:         {maxv:F4}");
        Console.WriteLine($"  Volatility:  {volatility:F2}%");
        Console.WriteLine($"  Current:     {last:F4}");
    }

    private async Task ChangePairAsync()
    {
        string b = Ask($"Base currency (default {baseCurrency}): ");
        string t = Ask($"Target currency (default {targetCurrency}): ");
        string newBase = string.IsNullOrEmpty(b) ? baseCurrency : b.ToUpper();
        string newTarget = string.IsNullOrEmpty(t) ? targetCurrency : t.ToUpper();
        if (newBase == baseCurrency && newTarget == targetCurrency)
        {
            Console.WriteLine("Pair unchanged.");
            return;
        }
        baseCurrency = newBase;
        targetCurrency = newTarget;
        var cached = cache.Get(baseCurrency, targetCurrency);
        if (cached != null && cached.Count > 0)
        {
            rates = cached;
            Console.WriteLine(C($"📂 Loaded cached data for {baseCurrency}/{targetCurrency}", Dim));
        }
        else
        {
            await FetchHistoryAsync();
        }
    }

    private async Task RefreshAsync()
    {
        await FetchHistoryAsync();
        Console.WriteLine(C("✅ Data refreshed.", Green));
    }

    private void ClearCache()
    {
        if (!AskConfirm("🗑️  Delete all cached data?")) return;
        cache.Clear();
        rates.Clear();
        Console.WriteLine(C("Cache cleared.", Yellow));
    }

    public async Task RunAsync()
    {
        Console.Clear();
